creusot/src/translation/specification.rs

```diff
@@ -32,7 +43,8 @@ impl<'tcx> PreContract<'tcx> {
 
     pub(crate) fn normalize(mut self, tcx: TyCtxt<'tcx>, param_env: ParamEnv<'tcx>) -> Self {
         for term in self.terms_mut() {
-            normalize(tcx, param_env, term);
+            *term =
+                normalize(tcx, param_env, std::mem::replace(term, /*Dummy*/ Term::mk_true(tcx)));
```

Comment:
Why change the signature like this? If you want it to be consisent wouldn't it be better to change `PreContract::normalize` instead?

Reply:
Yeah, I know.... My goal was to make these function normalize types in addition to resolving traits, but I failed because of `PartialEq::eq`. Given that rustc normalization consumes its argument, these function should as well.

I'm open to reverting this change. But the whole situation wrt when normalization happens is really bad, it is unclear when we have a type whether it is normalized or not (these functional invariants are not documented, and probably not maintained everywhere as well). I think there are two reasonable choices:

- If we decide to normalize eagerly, then this signature has to change, because `normalize` should normalize all the types, too. But we need to find a solution with `PartialEq::eq`.
- If we decide to normalize lazilly, then this is a rather big change. Hash tables of dependencies need to use type unification for testing equality, etc...

I think the last solution is better, but it requires more work.